compiler/hash-reporting/src/diagnostic.rs

```diff
@@ -2,50 +2,261 @@
-    /// store is implemented.
-    type DiagnosticsStore;
+use crate::reporter::Reports;
 
-    /// Get a reference to the associated [Self::DiagnosticsStore]
-    fn diagnostic_store(&self) -> &Self::DiagnosticsStore;
+macro_rules! make_diagnostic_traits {
+    ($($name:ident with $self_ref:ty),* $(,)?) => {
+        $(
+        pub trait $name<E, W>: Sized {
+            /// Add an error into the [Self]
+            fn add_error(self: $self_ref, error: E);
 
-    /// Add an error into the [Self]
-    fn add_error(&mut self, error: E);
+            /// Add an error from a [Result<T, E>] if the result is erroneous.
+            fn maybe_add_error<T>(self: $self_ref, value: Result<T, E>) {
+                if let Err(e) = value {
+                    self.add_error(e);
+                }
+            }
 
-    /// Add an error from a [Result<T, E>] if the result is erroneous.
-    fn maybe_add_error<T>(&mut self, value: Result<T, E>) {
-        if let Err(e) = value {
-            self.add_error(e);
+            /// Add a warning into the diagnostics store.
+            fn add_warning(self: $self_ref, warning: W);
+
+            /// Check if the diagnostics has an error.
+            fn has_errors(&self) -> bool;
+
+            /// Check if the diagnostics has a warning
+            fn has_warnings(&self) -> bool;
+
+            fn has_diagnostics(&self) -> bool {
+                self.has_errors() || self.has_warnings()
+            }
+
+            /// Convert the [Diagnostics] into a [`Vec<Report>`].
+            fn into_reports(
+                self: $self_ref,
+                make_reports_from_error: impl Fn(E) -> Reports,
+                make_reports_from_warning: impl Fn(W) -> Reports,
+            ) -> Reports {
+                let (errors, warnings) = self.into_diagnostics();
+                errors
+                    .into_iter()
+                    .flat_map(make_reports_from_error)
+                    .chain(warnings.into_iter().flat_map(make_reports_from_warning))
+                    .collect()
+            }
+
+            /// Convert the [Diagnostics] into its respective parts.
+            ///
+            /// This is useful when we just want to get the errors from the diagnostics rather than
+            /// immediately converting the diagnostics into [Report]s.
+            ///
+            /// This will modify self.
+            fn into_diagnostics(self: $self_ref) -> (Vec<E>, Vec<W>);
+
+            /// Merge another diagnostic store with this one.
+            fn merge_diagnostics(self: $self_ref, other: impl $name<E, W>);
+
+            /// Clear the diagnostics of all errors and warnings.
+            fn clear_diagnostics(self: $self_ref);
         }
+        )*
+    };
+}
+
+make_diagnostic_traits!(Diagnostics with &Self, DiagnosticsMut with &mut Self);
+
+pub struct ImmutableDiagnostics<E, W> {
+    pub errors: RefCell<Vec<E>>,
+    pub warnings: RefCell<Vec<W>>,
+}
```

Comment:
`@@Docs`, also would be nice to get rid of refcells somehow?

Reply:
Not really possible to get rid of refcells since the access is immutable..

Reply:
Oh right, the naming is a little confusing, as in it needs to be within an immutable context to append things to the diagnostic store.

Reply:
I'm open to better naming ideas

Reply:
Maybe something like `DiagnosticMut` to `Diagnostics`, and `ImmutableDiagnostics` to `CelledDiagnostics`?

Reply:
`DiagnosticStore` and `DiagnosticCellStore` ? Cause `Diagnostics` is the name of the trait